### PR TITLE
AMQNET-637 NMS 2.0 - Add NMS 2.0 method implementations.

### PR DESCRIPTION
**Pull request by michaelandrepearce**

### CODE REVIEW

**Comment by michaelandrepearce** [comment]
WIP - DO NOT MERGE

Still need to add tests and validate some bits

File: src/NMS.AMQP/NmsSession.cs [hunk 1]
```diff
@@ -396,8 +450,20 @@ public void AcknowledgeIndividual(AckType ackType, InboundMessageDispatch envelo
...
+        {
+            Task task = SendAsync(producer, destination, original, deliveryMode, priority, timeToLive, disableMessageId,
+                disableMessageTimestamp, deliveryDelay, completionListener);
+            if (completionListener == null)
+            {
+                task.ConfigureAwait(false).GetAwaiter().GetResult();
+            }
+        }
+
+        public Task SendAsync(NmsMessageProducer producer, IDestination destination, IMessage original, MsgDeliveryMode deliveryMode,
```
Comment (Havret) [inline src/NMS.AMQP/NmsSession.cs:465]: This is not the correct way of handling async sends. We should have either SendAsync that accepts callback and returns void or without Callback but returning Task. In this form it doesn't make sense.

**Comment by Havret** [comment]
Good beginning but there is some serious work to be done regarding async sends in `AmqpProducer`. Currently this is done on a fully synchronous way.

File: src/NMS.AMQP/NmsSession.cs (hunk 1 above)
Comment (michaelandrepearce) [reply src/NMS.AMQP/NmsSession.cs:465]: So it reduced duplicating alot of code. Also due to 2.0 api if someone sets a completion listener but still sends using async then the callback should be called. I dont expect people to do that but as its possible then need to handle it

**Comment by michaelandrepearce** [comment]
I dont fully follow, the actual send js being sent and is being done async within the taskbelow basically we are not blocking on waiting for the send task instead we pass tvaf back the task returned so its not blocking.
As when we discussed your original work it should be all async and simply the original 1.1 block send was meant to wait on the task, and all we do here is no longer back but pass it back.

**Comment by michaelandrepearce** [comment]
Lets discuss offline

File: src/NMS.AMQP/NmsSession.cs (hunk 1 above)
Comment (michaelandrepearce) [reply src/NMS.AMQP/NmsSession.cs:465]: That method isnt actually in api, its public though so the api methods use it, and can be used from the NMSProducer wrapper

**Comment by Havret** [comment]
I'm trying to do changes for `AmqpProducer` but I don't know what I should do for situation when
```
bool sync = deliveryMode == MsgDeliveryMode.Persistent;
```

Currently our `async` mode == fire and forget mode. I'm not sure how our new `SendAsync` bits fit into this picture.

**Comment by michaelpearce-gain** [comment]
So actually the logic should be for non-persistent messages, you can always do async even if sync send, because for non persistent there is no guaranteed delivery.

For persistent, on send sync, you need to block  for async you send async (Aka non block), but return task, and or return result on completion listener

**Comment by Havret** [comment]
Tbh I don't get it. 

I see 2 cases here. 

1) Sync fire and forget mode for non persistent messages. 
2) Async when we actually wait for the completion from the transport layer.

I wouldn't follow JMS strictly in terms of async api, as this concept essentially doesn't exist in Java. That's why I was against including passing callbacks, it totally obscures the api.

**Comment by michaelandrepearce** [comment]
Because for non persistent there is no need to block even at the packet send layer, as such you just send async and dont care for result. This was actually there in jms 1.1 for non persistent even...

For transaction even when persistent you can send async, you do need to track confirms async but then the transaction is sync, flushing any in flight async buffers etc.

**Comment by Havret** [comment]
> Because for non persistent there is no need to block even at the packet send layer, as such you just send async and dont care for result. This was actually there in jms 1.1 for non persistent even...

So this is sync. Async means that you actually can await the result.

**Comment by michaelpearce-gain** [comment]
So sync means its blocking on succesfull packet send, for non persistent and transaction you shouldn't do this, you async need to manage the repsonses only ...this is important not up at the .net layer but down at the packet tcp stack layers.

**Comment by michaelpearce-gain** [comment]
Please do go look at both Artemis and Qpid details around here....

**Comment by Havret** [comment]
https://softwareengineering.stackexchange.com/questions/393264/is-the-c-async-task-construct-equivalent-to-javas-executor-future

**Comment by michaelpearce-gain** [comment]
> https://softwareengineering.stackexchange.com/questions/393264/is-the-c-async-task-construct-equivalent-to-javas-executor-future

Im not saying this is anything to do with java vs .net its about semantics of the cases

**Comment by michaelpearce-gain** [comment]
@Havret  ok, removed the completionlistener, and just gone with task based async , using your new async send method.

It goes with obviously the removal of that in the api - https://github.com/apache/activemq-nms-api/pull/15 

Shall we merge this? and then iterate on any edge bits? Anything else you feel strongly on.
As well did you want to try get async receive added on the cosumer side, or shall we try get this out the door first?

**Comment by michaelpearce-gain** [comment]
@cjwmorgan-sol any comments?

**Comment by michaelpearce-gain** [comment]
note, pr is failing because nms api 2.0 isnt available yet, as the changes there havent been merged, and we'd need to release that.

**Comment by Havret** [comment]
@michaelandrepearce Check [this](https://github.com/michaelandrepearce/activemq-nms-amqp/pull/2) PR out. It should be fine now.

**Comment by cjwmorgan-sol** [comment]
The only concerns I have is for the order of execution of the async send and application network send vs send result notifications. 

This might just be my naiveté towards the C# async patterns. Wouldn't the async sends potentially get out of order as the task executor would be on a thread pool that is system default based? I do not see how the send tasks execution order is synchronized internal in the provider to ensure send order from the application thread. Perhaps I'm missing something? This might also be a concept for the completion result signal, sounds like above the async implementation might deviate a bit from JMS spec perhaps I'm missing some rationale behind that as well. Right now my view on a NMS async send would be the nio equivalent (or something close to) of the JMS 2.0 send.

Also the async send seems to produce different tasks depending on delivery mode (persistent vs non-persistent) having the task complete on network io complete for non-persistent and outcome result for persistent. This also leaves the application is a state where its unware of the messages send state. Although thinking about it with amqpnetlite "send" might mean buffered instead of io complete. However removing the capabilities of the application of separating network io complete vs response complete might be problematic back at the NMS api level should application require both network io complete and send outcome completion as separate events. Would such a  capability be out of scope for the NMS 2.0 improvements? Or is there some rationale I'm missing for the decide to combine the event of network io completion and send request completion?

I also tried the unit tests using the NMS 2.0 api from @michaelpearce-gain pull request they all passed. Are there any tests for the ordering of messages sent and received I'd image so I have not found them. (I'm pretty sure they should be somewhere).
I still need to try to update my testing tool a bit before I test any of the new significant changes done in NMS 2.0. Not sure when or if I'll be able to do that. I'll be relying heavily on testing in the meantime. :)

**Comment by Havret** [comment]
@cjwmorgan-sol So sends are fire and forget in AmqpNetLite. Actually all operations are fire and forget. You do not actually await any IO operation. Take a look at this --> https://github.com/Azure/amqpnetlite/issues/415. All you can await for in sync or async way is a reply from the broker itself. I'm not sure what it has to do with the order of messages, though.

**Comment by michaelandrepearce** [comment]
Non persistent does not need to wait for confirmation from broker. By spec there is no gaurentee of delivery.  Qpid actually does the same as well as Artemis jms client.

**Comment by michaelandrepearce** [comment]
Regarding async sent that and ordering for persistent will test that might be manual at first

**Comment by michaelandrepearce** [comment]
@Havret thanks merged that last bit. Quick question do we want to make receive async before we cut a release, like wise do we want add async create session producer etc. Or do we leave for 2.1?

**Comment by Havret** [comment]
bth first I would rather focus on adding tests to cover newly introduced features that come with 2.0 API.

**Comment by michaelandrepearce** [comment]
@Havret i will merge this: https://github.com/apache/activemq-nms-amqp/pull/58 , i had to amend the commit messages on the contributions you did as they didnt have the JIRA id prefix, as per guidelines

**Comment by michaelandrepearce** [comment]
fixed commit messages in branch, adding jira to those without in commit message

**Comment by modmynitro** [comment]
Hello, I am a bit confused, when can we expect a release of all this? I cant find any information on when NMS API 2.0 will be released.

**Comment by michaelpearce-gain** [comment]
Its still wip, its apache is all community driven, so it goes as fast as people can commit to. Contributors to help speed things up are always welcome, and is normally the best way to get what you need.

Obviously if you wanted to go down a more commercial route, there are a number of companies that provide support and technical services for activemq, clients etc, with contributors and committers on the project, you could always also engage with those if you need, a list of known ones are here: http://activemq.apache.org/support.html

File: test/Apache-NMS-AMQP-Test/TestAmqp/TestAmqpPeer.cs [hunk 2]
```diff
@@ -605,7 +606,49 @@ public void ExpectDurableSubscriberAttach(string topicName, string subscriptionN
...
+        {
+            Action<string> linkNameMatcher = linkName => Assert.AreEqual(subscriptionName+"|volatile1", linkName);
+
+            Action<object> sourceMatcher = o =>
+            {
+                Assert.IsInstanceOf<Source>(o);
+                var source = (Source) o;
+                Assert.AreEqual(topicName, source.Address);
+                Assert.IsFalse(source.Dynamic);
+                // Assert.AreEqual(TerminusDurability.UNSETTLED_STATE, (TerminusDurability) source.Durable);
```
Comment (michaelpearce-gain) [inline test/Apache-NMS-AMQP-Test/TestAmqp/TestAmqpPeer.cs:641]: @lukeabsent remove commented code?

**Comment by michaelandrepearce** [comment]
Closing , this work has now merged into: https://github.com/apache/activemq-nms-amqp/tree/2.0.x

**Pull request closed without merging**